### ISSUE TITLE
UWP support

### ISSUE THREAD

**Issue by vlasenkoalexey** [alanmcgovern/monotorrent#310]
I've been struggling with PlatformNotSupportedException that is thrown for some UWP clients, and stumbled on this: https://github.com/dotnet/platform-compat

I run analyzer, and it returned few interesting warnings:
MonoTorrent\MonoTorrent.Client.Connections\SocketConnection.cs(182,25,182,37): warning PC001: Socket.ConnectAsync(SocketAsyncEventArgs) isn't supported on Linux and macOS
MonoTorrent\MonoTorrent.Client.Listeners\PeerListener.cs(67,27,67,38): warning PC001: Socket.AcceptAsync(SocketAsyncEventArgs) isn't supported on Linux and macOS

Looks like .net standard it not that standard after all.

**Comment by alanmcgovern** [comment]
If you're hitting a PlatformNotSupported exception for UWP then please share as much info as you can so we can come up with a reasonable workaround.

I'm not sure how much I trust that analyzer though. The only thing it marks as potentially problematic are the Socket APIs, and they are supported on MacOS and Linux. If you're seeing a problem in UWP then this analyzer isn't catching the problem.

**Comment by vlasenkoalexey** [comment]
FYI MonoTorrent in not usable from UWP as is due to the way file system is implemented.
I had to rewrite File IO to make it work, but it is pretty ugly.

Regarding PlatformNotSupportedException, I still wasn't able to reproduce it, but believe that it is related to SHA1/MD5 providers.

**Comment by alanmcgovern** [comment]
I'd love to have it work out-of-box. If you can give me a sample showing the issue, and also share how you solved it, we should be able to find a  good solution we can build into the library!

**Comment by vlasenkoalexey** [comment]
The UWP implementation is not that straighforward, I could take it out of my project and share with you if you will be willing to make it work for UWP properly. I'm not sure that it is worth the effort since Microsoft is actively trying to kill it anyway.

To give you an idea, I had to reimplement FileStreamBuffer and DiskWriter to work using streams derived from StorageFile, and therefore make all file API async. And besides that replace all System.IO.File and System.IO.Directory API calls with replacement built on top of StorageFile/StorageFolder.

Let me know if you would be interested working on this.

**Comment by alanmcgovern** [comment]
> The UWP implementation is not that straighforward, I could take it out of my project and share with you if you will be willing to make it work for UWP properly. 

Yup! I think it's worth supporting it 'properly' and nuget should make it relatively easy to package up the result for external consumption :) I suspect uwp will be around for quite a while so I doubt it'd be wasted effort. 


> To give you an idea, I had to reimplement FileStreamBuffer and DiskWriter to work using streams derived from StorageFile, and therefore make all file API async. And besides that replace all System.IO.File and System.IO.Directory API calls with replacement built on top of StorageFile/StorageFolder.

I've made this code async I'm a branch, which will hopefully simplify further changes: https://github.com/alanmcgovern/monotorrent/commit/fb6be4d0066c02df5b5fb3896e7d3c2be61e8100 .

I'll double check those APIs and see if there's a common set which is supported in .NET Standard. If there isn't a common set I'm sure we can build in an easier way to work with alternative streams.

**Comment by alanmcgovern** [comment]
@vlasenkoalexey I made some changes to use UWP specific APIs within monotorrent itself. This is untested, but perhaps it'll help point the way towards how UWP could be supported in a reasonably clean way.

https://github.com/alanmcgovern/monotorrent/pull/313

**Comment by vlasenkoalexey** [comment]
This is great, I didn't think of retargeting project from .netstandard 2.0 to UWP, expected way more problems with that. If that's all what's needed, adding UWP support might be easier than I thought.

Here is the version of monotorrent I'm using: https://github.com/vlasenkoalexey/monotorrent/tree/uwp
Not all changes in this branch are related to UWP.

Main differences are:
- TorrentFileStream takes Stream instance in constructor and wraps it
- DiskWriter and FileStreamBuffer are reimplemented in MonoTorrent.UWP
- ClientEngine is created like this:

                MonoTorrent.HashAlgoFactory.SHA1Builder = () => System.Security.Cryptography.SHA1.Create();
                MonoTorrent.HashAlgoFactory.MD5Builder = () => System.Security.Cryptography.MD5.Create();
                engine = new ClientEngine(Globals.Settings, new MonoTorrent.UWP.Client.PieceWriters.DiskWriter());

- And there are bunch of other small changes related to replacing System.IO.File/ System.IO.Directory with MonoTorrent.UWP.File/MonoTorrent.UWP.Directory

Here is the app built using it: https://torrent-rt.com/
I'd prefer not to share app sources, at least for now. But I'm more than willing to help with properly implementing UWP support in MonoTorrent.

**Comment by alanmcgovern** [comment]
Awesome. Thanks for sharing those. I'll take a look through these changes and see what's easy to upstream :)

**Comment by alanmcgovern** [comment]
I took a brief look through the changes. There's a mix of things in there. I'd say 1/2 look good to clean up and merge into the primary repository, about 1/2 could be a little problematic or are already implemented :)

This is my super short summary:

I think we should replicate an NLog type experience internally and allow users of MonoTorrent to plug in actual NLog, or whatever they want, to capture the text logged by monotorrent :)

DiskManager/IPieceWriter has been made async throughout the core engine now.

Removing the code which randomises the list of trackers looks questionable (the spec asks for it to be randomised to help load balancing).

Removing the locking around PieceWriter usage looks like it could introduce race conditions. However since the core has already been made async we can *probably* dump these changes and rely on what's currently in the `unstable-vnext` branch :)

I'm curious why the 'proper' async method was replaced with a Task.Run and blocking synchronous code: https://github.com/vlasenkoalexey/monotorrent/commit/7c48291692093cad64039d9eedafe09efdf37605#diff-e75e50d9588383dd3d99fe27302979aaL573-L581

The changes to the Sha1 factory seem reasonable https://github.com/vlasenkoalexey/monotorrent/commit/7c48291692093cad64039d9eedafe09efdf37605#diff-7a288f25ae8532c52f2d60864e06889cR43-R44

This shouldn't be needed anymore https://github.com/vlasenkoalexey/monotorrent/commit/7c48291692093cad64039d9eedafe09efdf37605#diff-0b60a50b3f730e1f804f42516c4029eaR109-R110  and  https://github.com/vlasenkoalexey/monotorrent/commit/7c48291692093cad64039d9eedafe09efdf37605#diff-0b60a50b3f730e1f804f42516c4029eaR109-R110 . I fixed the issue where pieces could be written after-the-fact a few months ago.

This is pretty worrying. If you can't seek reliably then I'd be worried you wouldn't be able to read/write reliably too https://github.com/vlasenkoalexey/monotorrent/commit/7c48291692093cad64039d9eedafe09efdf37605#diff-82bcace67a81f124da6cbf22f08c52efR77

This will not cause anything to be run in parallel https://github.com/vlasenkoalexey/monotorrent/commit/7c48291692093cad64039d9eedafe09efdf37605#diff-c352b36b508b4b17b3769a23649e6003R62 . Also, this is super super fast anyway because it's (essentially) a loop which calls `Socket.Close`. It can't be made faster with parallelisation :) 

This looks like a no-op as the original code does the same as the new code? https://github.com/vlasenkoalexey/monotorrent/commit/7c48291692093cad64039d9eedafe09efdf37605#diff-48196dcc9cc1b587c932981bf75f3c75R568

It feels like a bug that this would be exposed to users of monotorrent https://github.com/vlasenkoalexey/monotorrent/commit/7c48291692093cad64039d9eedafe09efdf37605#diff-48196dcc9cc1b587c932981bf75f3c75R545-R551 . I wonder if this can be solved a better way?

We can make this configurable - https://github.com/vlasenkoalexey/monotorrent/commit/7c48291692093cad64039d9eedafe09efdf37605#diff-4a6a90c3e29e4bd8d3a55f87894054a1R35

The changes to FileStreamBuffer/DiskManager etc look like they can be rolled into monotorrent using conditional compilation for UWP :)

**Comment by alanmcgovern** [comment]
The question I have for you is - do you want to upstream these changes, basing your work on the `unstable-vnext` branch? If so I'll happily review, merge and release a new package with a uwp version of the library.

**Comment by vlasenkoalexey** [comment]
Thanks for detailed write-up. I didn't spend much time to make code clean, just hacked stuff to make it work for my case. So, sorry for mess :-)

Question about unstable-vnext, is it a branch for the upcoming release, or API breaking rewrite?
Sure take bits you feel are reusable and plug them into MonoTorrent as you see them fit.
I'll help with end-to-end testing. This will also make my work maintaining my branch much easier as well.

As for your questions:
Removing the locking around PieceWriter usage - actually I just removed it because async version doesn't support lock statement and forgot about it. 
If you think that it can lead to race conditions, then it makes sense to wrap it in async lock. I'm using one from Nitro.AsyncEx, but it should be possible to get implementation as part of MonoTorrent if you don't want to get dependency on other nuget packages. Probably it'll be necessary at some point as you replace more stuff with async.

Removing the code which randomises the list of trackers looks questionable (the spec asks for it to be randomised to help load balancing).
- you are right, this is no longer necessary. Before you did a rewrite for tracker logic to announce to all trackers in parallel it didn't work that well. From my experience usually torrents and magnet links put working trackers first in the list and bunch of others that might not necessarily work later. If you randomize list, you have lower chance that you'll hit working tracker.

Btw, I believe that there is still some issue with logic how discovered peers are added to the torrentmanager. Sometimes when you restart download, it takes a lot of time for peers to be added back. I didn't get to debug it further. You can play with Torrent RT to see what I'm talking about. Just find a torrent with multiple trackers.

I'm curious why the 'proper' async method was replaced with a Task.Run and blocking synchronous - honestly I don't remember, but believe that is due to the my early efforts to keep File IO API synchronous. Shouldn't be necessary.

This shouldn't be needed anymore vlasenkoalexey@7c48291#diff-0b60a50b3f730e1f804f42516c4029eaR109-R110 and vlasenkoalexey@7c48291#diff-0b60a50b3f730e1f804f42516c4029eaR109-R110 . I fixed the issue where pieces could be written after-the-fact a few months ago. - I'm not so sure, something like this might still be useful. The reason I added it here it to unblock delete file functionality. From my experience even after torrent is stopped disk manager still can keep writing queued blocks, and won't allow to delete file.

This is pretty worrying. If you can't seek reliably then I'd be worried you wouldn't be able to read/write reliably too - I think this is no longer necessary. As far as I remember I tried to use Async stream API without locking and it didn't work that well. Even with locking it wasn't stable.

The changes to the Sha1 factory seem reasonable vlasenkoalexey@7c48291#diff-7a288f25ae8532c52f2d60864e06889cR43-R44 - it should be easier to solve this using conditional compilation for UWP. And I'm still not sure if this is necessary, as I wasn't able to repro PlatformNotSupported exception on any of my machines.

This looks like a no-op as the original code does the same as the new code? vlasenkoalexey@7c48291#diff-48196dcc9cc1b587c932981bf75f3c75R568
- as far as I remember this is necessary because I rewrote the way ConnectionManager.CancelPendingConnects works. Please take a look, I think my implementation should work better. I'm using C++ UTP library, and therefore it has to be disposed carefully.  But I don't remember exactly why I had to do it this way. 

It feels like a bug that this would be exposed to users of monotorrent vlasenkoalexey@7c48291#diff-48196dcc9cc1b587c932981bf75f3c75R545-R551 . I wonder if this can be solved a better way? - could you explain why is this a bug?

We can make this configurable - vlasenkoalexey@7c48291#diff-4a6a90c3e29e4bd8d3a55f87894054a1R35 - agree, this should be configurable. On UWP whenever big file is created (few Gigs), create method can get blocked for few minutes on slower file systems. Having a bigger buffer to allow MonoTorrent to continue downloading during this time helps to mitigate this.

I guess that the biggest question remain is how to replace System.IO.File/Directory API calls.

**Comment by alanmcgovern** [comment]
Heya!

I'm just taking a look at UWP support again. Now that I've merged `unstable-vnext` branch into master everything that *can* be async *is* async. This should make it a little easier to support UWP specific behaviors via multi-targeting.

**Comment by alanmcgovern** [comment]
I think it works now? http://recordit.co/r2POaJaWk4

https://github.com/alanmcgovern/monotorrent/tree/uwp

Does that branch work for you? If not, what else needs to change?

**Comment by uponatime2019** [comment]
I can say that monotorrent worked for UWP without problem. I made a torrent downloader app with no issue

**Comment by vlasenkoalexey** [comment]
Sorry, got busy with other work. Changes look amazing, it is a nice improvement overall besides UWP. Will integrate them in next release and share results.

**Comment by alanmcgovern** [comment]
I'll close this for now as I don't understand what the issues are, or which solution would be best. I have some WIP branches I could never land as I didn't know if it was solving the right issues.

Would be more than happy to reopen the investigation if there's a clear set of deliverables which will resolve everything!
